### ISSUE TITLE
Codebook of `ManyClassClassifier` broken in `predict_proba` for large number of target classes

### ISSUE THREAD

**Issue by MaxSchambach** [PriorLabs/tabpfn-extensions#68]
### Describe the bug

The `ManyClassClassifier` extension seems to be broken for a large(-ish) number of classes, depending on some of its hyperparameters. I've observed this first on real data but was able to reproduce on synthetic data, see the code below for an example.

In the below example, the prediction breaks at `num_classes=87` for an alphabet size of 10 and a redundancy of 2. 

(The breaking pattern is somewhat strange. Depending on the data generation seed, it also occasionally breaks at 82 classes but works again for 83-86 classes. The breaking at 87 seems to be consistent across seeds though. So somehow, there is a dependency on the data as well which I don't get 🙃.)

The point of failure seems to depend on the alphabet size and chosen redundancy. (Larger breaking points for larger redundancy and alphabet sizes) but I didn't thoroughly ablate it and wasn't able to observe any obvious patterns here... 

E.g. using a redundancy of 4 in the below examples moves the point of failure to 222 classes and somewhere between 450 and 475 classes for a redundancy of 8.

The error appears in the assertion in [predict/predict_proba here](https://github.com/PriorLabs/tabpfn-extensions/blob/main/src/tabpfn_extensions/many_class/many_class_classifier.py#L311).
I'm not entirely sure what is happening there in detail. Up to the point of assertion things seem roughly fine as far as I can tell. E.g. the probability array produced has the correct dimensions so something seems to be off in the codebook. 

I've tried with both the latest version on PyPi (tagged as 0.0.4) as well as the latest on `main`.

### Steps/Code to Reproduce

Code example. The point after which it breaks somewhat depends on the alphabet size as well as the redundancy chosen. E.g. higher values of redundancy shifts the point a bit towards larger number of classes.

```python
import numpy as np
from tabpfn import TabPFNClassifier
from tabpfn_extensions.many_class import ManyClassClassifier
from tqdm import tqdm


def get_classifcation_data(num_classes: int, num_features: int, num_samples: int):
    assert num_samples > num_classes
    # Random feature matrix
    X = np.random.randn(num_samples, num_features)
    # Random target vector, making sure that every class appears at least once
    y = np.concatenate([np.arange(num_classes), np.random.randint(0, num_classes, size=num_samples-num_classes)])
    y = np.random.permutation(y)
    assert np.unique(y).size == num_classes
    return X, y


np.random.seed(42)
failures = []
for num_classes in tqdm(range(80, 100)):

    X, y = get_classifcation_data(
        num_classes=num_classes,
        num_features=4,
        num_samples=2*num_classes,
    )

    base_estimator = TabPFNClassifier(random_state=42)
    estimator = ManyClassClassifier(
        estimator=base_estimator,
        alphabet_size=10,
        n_estimators=None,  # auto-calculate number of estimators
        n_estimators_redundancy=2,  # default value: 4
        random_state=42,
    )
    try:
        estimator.fit(X, y)
        _ = estimator.predict(X)
    except Exception as e:
        print(f"Failed for: {num_classes}")
        failures.append((num_classes, e))

print("RESULTS:")
for failure in failures:
    print("-", failure)
```

### Expected Results

The expected result is that the many class extension works independently of the number of classes present in the training set and regardless of the chosen redundancy of the codebook.

### Actual Results

Prediction, `predict_proba` fails at different classes thresholds, depending on the alphabet size and redundancy chosen.

### Versions

```shell
PyTorch version: 2.5.1+cu124
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A
OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.28.0
Libc version: glibc-2.31
Python version: 3.11.11 (main, Dec 11 2024, 16:28:39) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-1073-azure-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.216.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        48 bits physical, 48 bits virtual
CPU(s):                               16
On-line CPU(s) list:                  0-15
Thread(s) per core:                   1
Core(s) per socket:                   16
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            AuthenticAMD
CPU family:                           23
Model:                                49
Model name:                           AMD EPYC 7V12 64-Core Processor
Stepping:                             0
CPU MHz:                              3209.125
BogoMIPS:                             4890.87
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            512 KiB
L1i cache:                            512 KiB
L2 cache:                             8 MiB
L3 cache:                             16 MiB
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; untrained return thunk; SMT disabled
Vulnerability Spec rstack overflow:   Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru arat umip rdpid
Dependency Versions:
--------------------
tabpfn: 2.0.8
torch: 2.5.1
numpy: 1.26.4
scipy: 1.15.2
pandas: 2.2.3
scikit-learn: 1.5.2
typing_extensions: 4.12.2
einops: 0.8.1
huggingface-hub: 0.29.3
```

**Comment by noahho** [comment]
Thanks a lot for the detailed error report! I will investigate.
At the same time we haven't built this version for high cardinality output classes and I don't expect this to work well with current code - we'd need to optimize more for it.
How costly would it be to try a few hyperparameters for optimization? I could point towards 1-2 things to look at if useful.

Will update shortly.

**Comment by MaxSchambach** [comment]
Thanks for the quick response!
I could definitely check out some hyperparameters combinations. However, ultimately, I am evaluating at scale with 400+ datasets with more than 10 and 50+ with more than 100 classes. So I would like to avoid per-dataset tuning 🙃

**Comment by MaxSchambach** [comment]
Hi @noahho, did you have some time have a look into it?

**Comment by noahho** [comment]
Dear @MaxSchambach , I've started looking into it and have something working. I might still take a couple of days however, as I can't prioritize work on the extensions repository currently. It will be ready latest next week.

**Comment by MaxSchambach** [comment]
Thank you! No worries, I understand and just wanted to check in :) Next week sounds great!
